File: src/go/problems/is_palindrome_number_test.go
// https://leetcode.com/problems/palindrome-number

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type IsPalindromeNumber struct{}
func TestIsPalindromeNumber(t *testing.T) {
	gen := core.TestSuite[IsPalindromeNumber]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(121).Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param(-121).Result(false)
	}).Add(func(tc *core.TestCase) {
		tc.Param(10).Result(false)
	}).Add(func(tc *core.TestCase) {
		tc.Param(1221).Result(true)
	}).Run(t)
}

func (IsPalindromeNumber) Solution1(x int) bool {
	if x < 0 {
		return false
	}

	original := x
	reversed := 0

	for x > 0 {
		reversed = reversed * 10 + x % 10
		x = x / 10
	}

	return original == reversed
}

func (IsPalindromeNumber) Solution2(x int) bool {
	if x < 0 {
		return false
	}

	arr := []int{}

	for x > 0 {
		arr = append(arr, x % 10)
		x = x / 10
	}

	left := 0
	right := len(arr) - 1
	for left <= right {
		if arr[left] != arr[right] {
			return false
		}

		left++
		right--
	}

	return true
}

File: src/python/problems/is_palindrome_number_test.py
#https://leetcode.com/problems/palindrome-number

from core.problem_base import *

class IsPalindromeNumber(ProblemBase):
    def Solution(self, x: int) -> bool:
        if x < 0:
            return False
        
        original = x
        reversed = 0
        while x > 0:
            reversed = reversed * 10 + x % 10
            x = x // 10
        
        return original == reversed                     
    
    def Solution1(self, x: int) -> bool:
        if x < 0:
            return False
        
        arr = []
        while x > 0:
            arr.append(x % 10)
            x = x // 10

        left = 0
        right = len(arr) - 1
        while left <= right:
            if arr[left] != arr[right]:
                return False

            left += 1
            right -= 1            
                
        return True

if __name__ == '__main__':
    TestGen(IsPalindromeNumber) \
        .Add(lambda tc: tc.Param(1221).Result(True)) \
        .Add(lambda tc: tc.Param(121).Result(True)) \
        .Add(lambda tc: tc.Param(-121).Result(False)) \
        .Add(lambda tc: tc.Param(10).Result(False)) \
        .Run()
